### PR TITLE
8 create resource entity dto repo service and controller

### LINKED ISSUE #8: Create Resource (Entity, DTO, Repo, Service, and Controller)
Resource ID
Name
Description
File Path (URL)
Uploaded By (User ID FK)
Category ID (FK)

### PR DESCRIPTION
**Pull request by darkhanzhursin**
- data access level
- uploads/downloads file

### CODE REVIEW

File: src/main/java/edu/miu/cs545/project/storage/StorageProperties.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("storage")
+public class StorageProperties {
+
+    /**
+     * Folder location for storing files
+     */
+    private String location = "upload-dir";
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/storage/StorageProperties.java]: extract to application.proprties

File: src/main/java/edu/miu/cs545/project/storage/StorageFileNotFoundException.java
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/storage/StorageFileNotFoundException.java:1]: create new package called exception

File: src/main/java/edu/miu/cs545/project/storage/FileSystemStorageService.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            }
+
+            if (fileName.contains("..")) {
+                throw new StorageException("Filename contains invalid path sequence " + fileName);
+            }
+            if (file.getBytes().length > (1024 * 1024)) {
+                throw new StorageException("File size exceeds maximum limit");
+            }
+            AcademicResource resource = new AcademicResource();
+            resource.setName(fileName);
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/storage/FileSystemStorageService.java]: set user by getting it as a parameter for store method

File: src/main/java/edu/miu/cs545/project/model/entity/AcademicResource.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import jakarta.persistence.*;
+import lombok.Data;
+
+@Entity
+@Data
+public class AcademicResource {
+
+    @Id
+    @GeneratedValue
+    private Long id;
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/model/entity/AcademicResource.java:13]: @Setter(AccessLevel.NONE)

File: src/main/java/edu/miu/cs545/project/controller/FileUploadController.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+@RequestMapping("/api/v1/resource")
+@Tag(name = "Resources", description = "Resource API")
+@RequiredArgsConstructor
+public class FileUploadController {
+
+    private final StorageService storageService;
+
+    // Looks up the current list of uploaded files
+    @GetMapping("/")
+    public String listUploadedFiles(Model model) throws IOException {
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/controller/FileUploadController.java]: remove

File: src/main/java/edu/miu/cs545/project/ProjectApplication.java [hunk 5]
```diff
@@ -1,15 +1,28 @@
...
 public class ProjectApplication {
 
     public static void main(String[] args) {
         SpringApplication.run(ProjectApplication.class, args);
     }
 
+    @Bean
+    CommandLineRunner init(StorageService storageService) {
+        return args -> {
+            storageService.deleteAll();
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/ProjectApplication.java]: remove

File: src/main/java/edu/miu/cs545/project/service/StorageService.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.web.multipart.MultipartFile;
+
+import java.nio.file.Path;
+import java.util.stream.Stream;
+
+public interface StorageService {
+
+    void init();
+    void store(MultipartFile file, Long userId);
+    Stream<Path> loadAll();
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/service/StorageService.java]: we don't need loadAll

File: src/main/java/edu/miu/cs545/project/service/StorageService.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    Path load(String filename);
+    Resource loadAsResource(String filename);
+    void deleteAll();
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/service/StorageService.java]:
```suggestion
    void delete(String filename);
```

File: src/main/java/edu/miu/cs545/project/controller/FileUploadController.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        Resource file = storageService.loadAsResource(filename);
+
+        if (file == null)
+            return ResponseEntity.notFound().build();
+
+        return ResponseEntity.ok().header(HttpHeaders.CONTENT_DISPOSITION,
+                "attachment; filename=\"" + file.getFilename() + "\"").body(file);
+    }
+
+//    @PostMapping("/upload")
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/controller/FileUploadController.java]: let's remove comments

File: src/main/java/edu/miu/cs545/project/controller/FileUploadController.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        return "redirect:/";
+    }
+
+    @ExceptionHandler(StorageFileNotFoundException.class)
+    public ResponseEntity<?> handleStorageFileNotFound(StorageFileNotFoundException exc) {
+        return ResponseEntity.notFound().build();
+    }
+
+//    @PostMapping("/single/upload")
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/controller/FileUploadController.java]: let's remove comments

**Review by BtsuanEyakem — APPROVED** [review]

File: src/main/java/edu/miu/cs545/project/exception/StorageFileNotFoundException.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package edu.miu.cs545.project.exception;
+
+public class StorageFileNotFoundException extends StorageException {
```
Comment (nabil-miu) [inline src/main/java/edu/miu/cs545/project/exception/StorageFileNotFoundException.java:3]: can this extends  RuntimeException directly?

**Review by nabil-miu — APPROVED** [review]

File: src/main/java/edu/miu/cs545/project/exception/StorageFileNotFoundException.java (hunk 10 above)
Comment (darkhanzhursin) [reply src/main/java/edu/miu/cs545/project/exception/StorageFileNotFoundException.java:3]: Yes, StorageFileNotFoundException can extend directly from RuntimeException. This depends on how you want to structure your exceptions and what kind of behavior you want them to have. If you want StorageFileNotFoundException to be a more specific type of StorageException, it makes sense to extend from StorageException. However, if you want it to be a standalone runtime exception, you can extend directly from RuntimeException

Comment (nabil-miu) [reply src/main/java/edu/miu/cs545/project/exception/StorageFileNotFoundException.java:3]: i am OK with both options

**Pull request merged**